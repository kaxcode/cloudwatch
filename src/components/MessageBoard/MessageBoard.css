.MessageBoard__Input {
  grid-area: message;
  padding: 30px 30px 0 30px;
}

.MessageBoard__Output {
  position: absolute;
  z-index: 1000;
  margin: 0 auto;
  text-align: center;
  font-family: "Baloo Bhai", cursive;
  background-color: grey;
}
