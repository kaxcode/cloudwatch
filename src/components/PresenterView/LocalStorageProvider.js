import React from 'react';
import { arrayOf, string, func } from 'prop-types';
import WindowEvent from './WindowEvent.js';

class LocalStorageProvider extends React.Component {
  static defaultProps = {
    children: () => {}
  }

  static propTypes = {
    keys: arrayOf(string),
    children: func.isRequired
  }

  state = {}

  componentWillMount = () => {
    this.deriveStateFromStore();
  }

  deriveStateFromStore = () => {
    const newState = this.props.keys.reduce((set, key) => {
      set[key] = localStorage.getItem(key);
      return set;
    }, {});
    this.setState(newState);
  }

  handleStorage = () => {
    this.deriveStateFromStore();
  }

  render() {
    return (
      <React.Fragment>
        <WindowEvent event="storage" handler={this.handleStorage} />
        {this.props.children(this.state)}
      </React.Fragment>
    );
  }
}

export default LocalStorageProvider;
