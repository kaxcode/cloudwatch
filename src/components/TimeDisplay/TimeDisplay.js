import { millisecondsToHuman } from '../../utils/humanizeTimer';
import { number, oneOfType, string } from 'prop-types';
import React from 'react';

const TimeDisplay = ({time}) => (
  <h3 className="Stopwatch__Count">
    {millisecondsToHuman(time)}
  </h3>
);

TimeDisplay.propTypes = {
  time: oneOfType([number, string])
};

export default TimeDisplay;
