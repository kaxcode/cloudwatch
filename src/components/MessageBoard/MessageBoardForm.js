import React from 'react';
import { func } from 'prop-types';
import './MessageBoard.css';

const MessageOutput = props => {
  if (window.name === 'presenter') {
    return null;
  }

  return (
    <form id="message-submit-form" onSubmit={props.onSubmit}>
      <input id="message-input" onChange={props.onChange} />
      <button id="message-submit-button">Post Message</button>
      <button onClick={props.onPresent}>Open Presentation Window</button>
    </form>
  );
};

MessageOutput.propTypes = {
  onChange: func.isRequired,
  onSubmit: func.isRequired,
  onPresent: func.isRequired
};

export default MessageOutput;
