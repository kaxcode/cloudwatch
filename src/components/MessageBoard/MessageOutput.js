import React from 'react';
import { func } from 'prop-types';
import './MessageBoard.css';
import PropTypes from 'prop-types';

const MessageOutput = props => {
  if(!props.message) {
    return null;
  }
  return (
    <div className="MessageBoard__Output">
      <h1>{props.message}</h1>
      <button onClick={props.onDismiss}>X</button>
    </div>
  );
};

MessageOutput.propTypes = {
  onDismiss: func.isRequired,
  message: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
};

export default MessageOutput;
