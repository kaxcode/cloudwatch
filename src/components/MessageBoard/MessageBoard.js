import { bool, func } from 'prop-types';
import React from 'react';
import MessageOutput from './MessageOutput';
import './MessageBoard.css';
import MessageBoardForm from './MessageBoardForm';

export default class MessageBoard extends React.Component {
  static defaultProps = {
    showInput: true,
    windowOpener: window.open.bind(window)
  };

  static propTypes = {
    showInput: bool,
    windowOpener: func.isRequired
  };

  state = {
    message: ''
  };

  componentDidMount = () => {
    if (window.name === 'presenter') {
      this.setState({ message: localStorage.message });
    }
  };

  handleDismiss = () => {
    this.setState({ message: '' });
    localStorage.setItem('message', '');
  };

  handleSubmit = (e, message) => {
    e.preventDefault();
    this.setState({ message });
    localStorage.setItem('message', message);
  };

  handlePresent = () => {
    this.props.windowOpener('present', 'presenter');
  };

  render() {
    return (
      <div className="MessageBoard__Input">
        <MessageBoardForm
          onSubmit={this.handleSubmit}
          onPresent={this.handlePresent}
        />
        <MessageOutput
          message={localStorage.message}
          onDismiss={this.handleDismiss}
        />
      </div>
    );
  }
}
