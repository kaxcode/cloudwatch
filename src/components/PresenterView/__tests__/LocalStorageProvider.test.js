import React from 'react';
import Enzyme from 'enzyme';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import LocalStorageProvider from '../LocalStorageProvider';
import MessageBoard from '../../MessageBoard/MessageBoard';
import TimeDisplay from '../../TimeDisplay/TimeDisplay';
import 'jest-localstorage-mock';

Enzyme.configure({ adapter: new Adapter() });

describe('LocalStorageProvider', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <LocalStorageProvider keys={['timeRemaining']}>
        {({ timeRemaining }) => (
          <React.Fragment>
            <TimeDisplay time={timeRemaining} />
            <MessageBoard />
          </React.Fragment>
        )}
      </LocalStorageProvider>
    );
  });

  afterEach(() => {
    localStorage.clear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('#componentDidMount', () => {
  it('sets the state of timeRemaining to that in localStorage', () => {
    // Arrange
    const wrapper = shallow(
      <LocalStorageProvider keys={['timeRemaining']}>
        {({ timeRemaining }) => (
          <React.Fragment>
            <TimeDisplay time={timeRemaining} />
            <MessageBoard />
          </React.Fragment>
        )}
      </LocalStorageProvider>
    );
    localStorage.setItem('timeRemaining', 10);
    // Act
    wrapper.instance().componentDidMount();
    // Assert
    expect(wrapper.state().timeRemaining).toEqual('10');
  });
});
describe('#handleStorage', () => {
  it('formats and sets the state to that in localStorage', () => {
    // Arrange
    const wrapper = shallow(
      <LocalStorageProvider keys={['timeRemaining']}>
        {({ timeRemaining }) => (
          <React.Fragment>
            <TimeDisplay time={timeRemaining} />
            <MessageBoard />
          </React.Fragment>
        )}
      </LocalStorageProvider>
    );
    localStorage.setItem('timeRemaining', 10);
    // Act
    wrapper.instance().handleStorage();
    // Assert
    expect(wrapper.state().someState.timeRemaining).toEqual('10');
  });
});
