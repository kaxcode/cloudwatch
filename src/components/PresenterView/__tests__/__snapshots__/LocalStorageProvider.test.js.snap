// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalStorageProvider renders correctly 1`] = `
<React.Fragment>
  <WindowEvent
    event="storage"
    handler={[Function]}
  />
  <div>
    456
  </div>
</React.Fragment>
`;
